pkg/codegen/go/gen.go

```diff
@@ -468,6 +508,100 @@ func genInputMethods(w io.Writer, name, receiverType, elementType string, ptrMet
 	}
 }
 
+func (pkg *pkgContext) genEnum(w io.Writer, enum *schema.EnumType) {
+	pkg.genEnumType(w, pkg.tokenToEnum(enum.Token), enum)
+}
+
+func (pkg *pkgContext) genEnumType(w io.Writer, name string, enumType *schema.EnumType) {
+	mod := pkg.tokenToPackage(enumType.Token)
+	modPkg, ok := pkg.packages[mod]
+	contract.Assert(ok)
+	printCommentWithDeprecationMessage(w, enumType.Comment, "", false)
+	elementType := pkg.enumElementType(enumType.ElementType, false)
+	fmt.Fprintf(w, "type %s %s\n\n", name, elementType)
+
+	fmt.Fprintln(w, "const (")
+	for _, e := range enumType.Elements {
+		printCommentWithDeprecationMessage(w, e.Comment, e.DeprecationMessage, true)
+		var elementName = e.Name
+		// Add the type to the name to disambiguate constants used for enum values
+		if e.Name == "" {
+			elementName = fmt.Sprintf("%v", e.Value)
+		}
+		enumName := makeSafeEnumName(elementName)
+		if strings.Contains(enumName, "_") {
+			enumName = fmt.Sprintf("_%s", enumName)
+		}
+		e.Name = name + enumName
+		contract.Assertf(!modPkg.names.has(e.Name), "Name collision for enum constant: %s for %s",
+			e.Name, enumType.Token)
+		switch reflect.TypeOf(e.Value).Kind() {
+		case reflect.String:
+			fmt.Fprintf(w, "%s = %s(\"%v\")\n", e.Name, name, e.Value)
+		default:
+			fmt.Fprintf(w, "%s = %s(%v)\n", e.Name, name, e.Value)
+		}
+	}
+	fmt.Fprintln(w, ")")
+	inputType := pkg.inputType(enumType, false)
+	contract.Assertf(name == inputType,
+		"expect inputType (%s) for enums to be the same as enum type (%s)", inputType, enumType)
+	pkg.genEnumInputFuncs(w, name, enumType, elementType, inputType)
+}
+
+func (pkg *pkgContext) enumElementType(t schema.Type, optional bool) string {
+	suffix := ""
+	if optional {
+		suffix = "Ptr"
+	}
+	switch t {
+	case schema.BoolType:
+		return "pulumi.Bool" + suffix
+	case schema.IntType:
+		return "pulumi.Int" + suffix
+	case schema.NumberType:
+		return "pulumi.Float64" + suffix
+	case schema.StringType:
+		return "pulumi.String" + suffix
+	default:
+		// We only expect to support the above element types for enums
+		panic(t)
+	}
+}
+
+func makeSafeEnumName(name string) string {
+	return makeValidIdentifier(Title(name))
```

Comment:
We had discussed bringing in the [common enum replacements](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/utilities.go#L124-L129) here... is that still the plan?

Reply:
Yes, sorry missed that. Updated.

Reply:
Might also be worth adding a [check](https://github.com/pulumi/pulumi/blob/41a881a37418291b3278ba32e62b5fbd4e3a8af0/pkg/codegen/nodejs/utilities.go#L110-L113) for names that are one illegal character.